File: python/repl/basic/parser.py
from ast import LetStatement
from token_type import TokenType


class Parser(object):
    def __init__(self, lexer):
        self.lexer = lexer
        self.current_token = None
        self.peek_token = None

    def reset(self, source):
        self.lexer.new(source=source)
        self.next_token()
        self.next_token()

    def register_prefix(self):
        pass

    def register_inflix(self):
        pass

    def next_token(self):
        self.current_token = self.peek_token
        self.peek_token = self.lexer.next_token()

    def current_token_is(self, token_type):
        return self.current_token.token_type == token_type

    def peek_token_is(self, token_type):
        return self.peek_token.token_type == token_type

    def expect_peek(self, token_type):
        if self.peek_token_is(token_type):
            self.next_token()
            return True
        return False

    def build_program_statements(self, program):
        # this returns a program based on the source of the lexer, this is better naming for this method
        # this is the old ParseProgram functions
        while self.current_token.token_type != TokenType.EOF:
            statement = self.parse_statement()
            if statement:
                program.statements.append(statement)
            self.next_token()

    def parse_statement(self):
        if self.current_token.token_type == TokenType.LET:
            return self.parse_let_statement()
        if self.current_token.token_type == TokenType.RETURN:
            return self.parse_return_statement()
        if self.current_token.token_type == TokenType.WHILE:
            return self.parse_while_statement()
        return self.parse_expression_statement()

    def parse_let_statement(self):
        let_statement = LetStatement()
        #print let_statement
        return self.current_token

    def parse_return_statement(self):
        return self.current_token

    def parse_while_statement(self):
        # this might be different in the final product
        return self.current_token

    def parse_expression_statement(self):
        return self.current_token

    def parse_expression(self, precedence):
        pass

    def parse_identifier(self):
        pass

    def parse_integer_literal(self):
        pass

    def parse_prefix_expression(self):
        pass

    def peek_precedence(self):
        pass

    def current_precedence(self):
        pass

    def parse_inflix_expression(self, left):
        pass


class ParsingPrecendece(object):
    LOWEST = 1
    EQUALS = 2
    LESSGREATER = 3
    SUM = 4
    PRODUCT = 5
    PREFIX = 6
    CALL = 7

File: java/VmImplementation2/src/parser/Parser.java
package parser;

import ast.AstNode;
import ast.nodes.*;
import lexer.Lexer;
import token.Token;
import token.TokenType;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

public class Parser {
    private Lexer lexer;
    private Token currentToken;
    private Token peekToken;
    private String source;
    private Map<TokenType, Supplier<AstNode>> prefixParseFns = new HashMap<>();
    private Map<TokenType, Function<AstNode, AstNode>> infixParseFns = new HashMap<>();
    private Map<TokenType, PrecedenceType> precedenceMap = new HashMap<>();

    public Parser() {
        lexer = new Lexer();
        prefixParseFns.put(TokenType.IDENT, this::parseIdentifier);
        prefixParseFns.put(TokenType.MINUS, this::parsePrefixExpression);
        prefixParseFns.put(TokenType.INT, this::parseIntegerLiteral);
        infixParseFns.put(TokenType.PLUS, this::parseInfixExpression);
        infixParseFns.put(TokenType.ASSIGN, this::parseAssignExpression);
        constructPrecedences();
    }

    private void constructPrecedences() {
        precedenceMap.put(TokenType.PLUS, PrecedenceType.SUM);
        precedenceMap.put(TokenType.MINUS, PrecedenceType.SUM);
        precedenceMap.put(TokenType.MULTIPLY, PrecedenceType.MULTIPLY);
        precedenceMap.put(TokenType.DIVIDE, PrecedenceType.MULTIPLY);
        precedenceMap.put(TokenType.ASSIGN, PrecedenceType.ASSIGN);
    }

    public void load(String source) {
        this.source = source;
        lexer.load(source);
        nextToken();
        nextToken();
    }

    private void nextToken() {
        currentToken = peekToken;
        peekToken = lexer.nextToken();
    }

    public void buildStatements(AstProgram program) {
        while (currentToken.getType() != TokenType.EOF) {
            AstNode statement = parseStatement();
            if (statement != null) {
                //System.out.println(statement);
                program.getStatements().add(statement);
            }
            nextToken();
        }
    }

    private AstNode parseStatement() {
        // if let statement return parseLetStatement
        // if return statement return parseReturnStatement
        return parseExpressionStatement();
    }

    private AstNode parseExpressionStatement() {
        AstExpressionStatement stmt = new AstExpressionStatement(currentToken);
        stmt.setExpression(parseExpression(PrecedenceType.LOWEST));
        if (peekTokenIs(TokenType.SEMICOLON)) {
            nextToken();
        }
        return stmt;
    }

    private AstNode parseExpression(PrecedenceType precedence) {
        if (!prefixParseFns.containsKey(currentToken.getType())) {
            return null;
        }
        Supplier<AstNode> prefix = prefixParseFns.get(currentToken.getType());
        AstNode leftExpression = prefix.get();
        while (!peekTokenIs(TokenType.SEMICOLON) && precedence.ordinal() < peekPrecedence().ordinal()) {
            if (!infixParseFns.containsKey(peekToken.getType())) return leftExpression;
            Function<AstNode, AstNode> infix = infixParseFns.get(peekToken.getType());
            nextToken();
            leftExpression = infix.apply(leftExpression);
        }
        return leftExpression;
    }

    private boolean currentTokenIs(TokenType type) {
        return currentToken.getType() == type;
    }

    private boolean peekTokenIs(TokenType type) {
        return peekToken.getType() == type;
    }

    private PrecedenceType currentPrecedence() {
        if (precedenceMap.containsKey(currentToken.getType())) return precedenceMap.get(currentToken.getType());
        return PrecedenceType.LOWEST;
    }

    private PrecedenceType peekPrecedence() {
        if (precedenceMap.containsKey(peekToken.getType())) return precedenceMap.get(peekToken.getType());
        return PrecedenceType.LOWEST;
    }

    private AstNode parseIdentifier() {
        AstIdent ident = new AstIdent(currentToken, currentToken.getLiteral());
        return ident;
    }

    private AstNode parsePrefixExpression() {
        AstPrefixExpression expr = new AstPrefixExpression(currentToken);
        nextToken();
        expr.setRight(parseExpression(PrecedenceType.PREFIX));
        return expr;
    }

    private AstNode parseInfixExpression(AstNode left) {
        AstInfixExpression expr = new AstInfixExpression(currentToken, currentToken.getLiteral(), left);
        PrecedenceType precedence = currentPrecedence();
        nextToken();
        expr.setRight(parseExpression(precedence));
        return expr;
    }

    private AstNode parseIntegerLiteral() {
        AstInteger node = new AstInteger(currentToken, Integer.valueOf(currentToken.getLiteral()));
        return node;
    }

    private AstNode parseAssignExpression(AstNode left) {
        AstAssignExpression stmt = new AstAssignExpression(currentToken);
        stmt.setName(left);
        nextToken();
        AstNode right = parseExpression(PrecedenceType.LOWEST);
        stmt.setValue(right);
        if (currentTokenIs(TokenType.SEMICOLON)) {
            nextToken();
        }
        return stmt;
    }
}
